python/lsst/afw/typehandling/_HeteroMap.cc

```diff
@@ -0,0 +1,188 @@
+/*
+ * This file is part of afw.
+ *
+ * Developed for the LSST Data Management System.
+ * This product includes software developed by the LSST Project
+ * (https://www.lsst.org).
+ * See the COPYRIGHT file at the top-level directory of this distribution
+ * for details of code ownership.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "pybind11/pybind11.h"
+#include "pybind11/stl.h"
+
+#include <exception>
+#include <iostream>
+#include <utility>
+
+#include "lsst/utils/python.h"
+#include "lsst/pex/exceptions.h"
+
+#include "lsst/afw/typehandling/HeteroMap.h"
+
+namespace py = pybind11;
+using namespace py::literals;
+
+namespace lsst {
+namespace afw {
+namespace typehandling {
+
+namespace {
+
+// Type safety pointless in Python, use unsafe methods to avoid manual type checking
+// https://pybind11.readthedocs.io/en/stable/advanced/classes.html#binding-protected-member-functions
+template <typename K>
+class Publicist : public MutableHeteroMap<K> {
+public:
+    using MutableHeteroMap<K>::unsafeLookup;
+    using MutableHeteroMap<K>::unsafeErase;
+};
+
+// This class lets us implement __getitem__ using HeteroMap::unsafeLookup
+// Nonlocal class definition to let us use a method template for well-behaved types
+class VariantOneWayCaster {
+public:
+    template <typename T>
+    py::object operator()(T const& value) {
+        return py::cast(value);
+    }
+    // HeteroMap::get returns the same reference, depend on RVP to avoid double-deletion
+    py::object operator()(std::unique_ptr<Storable> const& pointer) { return py::cast(*pointer); }
```

Comment:
I think I only used `reference_internal` to conform to the style guide. I *think* we could use `return_value_policy::copy` with no loss of functionality -- we don't need it for Python-side polymorphism, or for `__setitem__`. The only down side is if the copy is expensive.

Reply:
Hmm... we can't use `copy` because you should be able to modify mutable `Storable` objects from Python and have the map reflect that. I don't like forcing C++ users to hold everything by `shared_ptr`, especially because of potential aliasing problems, but that might be the least of the evils here.

Reply:
I do think it comes down to copy (which would actually need to invoke `clone`, of course) vs only supporting `shared_ptr` to `Storable`.  This is the kind of fundamental mismatch between Python and C++ that makes me really want to strongly encourage all `Storables` to be immutable.

Reply:
You know, I'm not sure this is a problem. You say it's a mismatch between Python and C++, but you could follow the exact same steps in C++ (including for standard library containers, which return deletable elements by reference). If we can't defend against such code in C++, I'm not sure there's much point in worrying about it in Python, either.

Reply:
I think the difference is that undefined behavior is something C++ developers are familiar with and (should!) always guard against.  Python developers have the expectation that the worst they can do is cause a runtime exception to be raised.

Reply:
The only current elements of `ExposureInfo` that are modifiable from Python are `PropertySet` [metadata], `Psf` and `ApCorrMap`; of these, `Psf`'s only Python modifier is `setCacheCapacity`. We might actually be able to get away with `copy`.

Reply:
I think I'll leave the behavior as-is. We have existing code that modifies metadata, and we have existing code that gets/sets `Filter` by value. We can't make the issue you described impossible unless we break one of those two use cases, and any legitimate desire to hang onto an object after deleting it could be handled naturally by `MutableMapping.pop`/`MutableMapping.popitem`.

Reply:
FWIW, metadata (like all of those other components) is currently held by `shared_ptr`, and I hadn't (until just now) thought we'd even consider changing that.  I suppose it may be most worth exploring not using `shared_ptr` for precisely the components that are mutable.

Reply:
Oops, you're right. `copy` really is safe, at least with current usage.

I'm confused by that last sentence -- `shared_ptr` gives us reference semantics without the risk of dangling references, so why would we specifically drop it for mutable components, where reference semantics are most useful?

Reply:
Just because mutability is what makes aliasing bad - but that's definitely a minor concern compared to dangling references, especially as it's not a new concern for us.

Reply:
Hmm... the [style guide](https://developer.lsst.io/coding/intro.html#stringency-levels) says: 
> MUST and SHALL
    ... The developer needs to petition the lead developer to acquire explicit approval to contravene the Rule.

Who's the lead developer?

Reply:
@ktlim, I suppose? I wasn't really familiar with that language either.

Reply:
@ktlim, we'd like an exception to the rule
> The reference_internal policy SHALL be used for functions (or properties) giving write access to internal data members

In this case, we're returning a non-`const` reference to a map element (consistent with idiomatic C++), but we're worried that might lead to broken references if somebody calls the pybind11 getter, saves the result, and then deletes the element from the map. We wish to use the `copy` policy instead.

We're not worried about being able to modify the values where this bug could come up (in particular, it would only happen if the element was originally added to the map from the C++ side).

Reply:
Yes, the code in question got refactored out after @TallJimbo did his review, but the underlying issue is with how the getters behave.

Reply:
I've appended the following text to the (in-Python) type explanation for `GenericMap`, `MutableGenericMap`, and `SimpleGenericMap`:
> As a safety precaution, `~lsst.afw.typehandling.Storable` objects that are
added from C++ may be copied when you retrieve them from Python, making it
impossible to modify them in-place. This issue does not affect objects that
are added from Python, or objects that are always passed by
:cpp:class:`shared_ptr` in C++.